src/main/java/org/prebid/server/bidder/loyal/LoyalBidder.java

```diff
@@ -0,0 +1,171 @@
+package org.prebid.server.bidder.loyal;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.Imp;
+import com.iab.openrtb.response.Bid;
+import com.iab.openrtb.response.BidResponse;
+import com.iab.openrtb.response.SeatBid;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.loyal.proto.LoyalImpExt;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderCall;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
+import org.prebid.server.proto.openrtb.ext.request.loyal.ExtImpLoyal;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.BidderUtil;
+import org.prebid.server.util.HttpUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+public class LoyalBidder implements Bidder<BidRequest> {
+
+    private static final TypeReference<ExtPrebid<?, ExtImpLoyal>> LOYAL_EXT_TYPE_REFERENCE =
+            new TypeReference<>() {
+            };
+
+    private static final String PUBLISHER_PROPERTY = "publisher";
+    private static final String NETWORK_PROPERTY = "network";
+    private static final String BIDDER_PROPERTY = "bidder";
+
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public LoyalBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<HttpRequest<BidRequest>> httpRequests = new ArrayList<>();
+
+        for (Imp imp : request.getImp()) {
+            try {
+                final ExtImpLoyal extImpLoyal = parseExtImp(imp);
+                if (isValidImp(extImpLoyal)) {
```

Comment:
no sure we need this check, one of those field are required by the json scheme

Reply:
You said:
[AntoxaAntoxic](https://github.com/AntoxaAntoxic) [2 weeks ago](https://github.com/prebid/prebid-server-java/pull/3183#discussion_r1605199217)
the following logic is missing

	if len(adapterRequests) == 0 {
		errs = append(errs, errors.New("found no valid impressions"))
		return nil, errs
	}


and this is part of this

Reply:
so this is the Go code
```
if loyalExt.PlacementID != "" {
			impExt.LoyalBidderExt.PlacementID = loyalExt.PlacementID
			impExt.LoyalBidderExt.Type = "publisher"
		} else if loyalExt.EndpointID != "" {
			impExt.LoyalBidderExt.EndpointID = loyalExt.EndpointID
			impExt.LoyalBidderExt.Type = "network"
		}
```

as you see the imp that doesn't have the `placementId` and `endpointId` fields won't be treated as invalid, but you've added the unit test that defines invalid bid as the following:

```
private static Imp givenBadImp(UnaryOperator<Imp.ImpBuilder> impCustomizer) {
        return impCustomizer.apply(Imp.builder()
                        .id("invalidImp")
                        .ext(mapper.valueToTree(ExtPrebid.of(null,
                                ExtImpLoyal.of(null, null)))))
                .build();
    }
```

So my conclusion the logic of your version of the bidder differs from the GO's version

Reply:
okey, now its clear. I hade made wrong badImp, I was doing it in wrong way.